### ISSUE TITLE
Implement Tread Cycling

### ISSUE THREAD

**Issue by Nostrademous** [Nostrademous/Dota2-FullOverwrite#125]
As appropriate for action being taken:

Retreating: Strength
Fighting: Agility
Casting: Intelligence

**Comment by pbenologa** [comment]
Any idea what GetPowerTreadsStat( ) returns exactly? It says int but how do i know if the stat is Strength, Agility or Intelligence?

**Comment by Nostrademous** [comment]
I believe it is:
0 = Str, 1 = Agi, 2 = Intelligence

**Comment by pbenologa** [comment]
I see. Thanks! I'll try making a function for this

**Comment by Nostrademous** [comment]
I will check this soon. have to fix entire codebase for new patch changes.

**Comment by pbenologa** [comment]
Seems like the code for checking enemy data is affected big time. In the meantime ill try making a secret shop buy for courier

**Comment by Nostrademous** [comment]
I implemented this, haven't checked if it works. 

BY THE WAY... to check things faster... in a lobby game enable cheats.

Then in console type:
sv_cheats 1

Now you can type in console:

host_timescale 5.0

to make time and all actions go 5 times faster... set back to 1.0 when you want to see things at normal speed.

**Comment by pbenologa** [comment]
Yup I usually do this. I do this when simulating a bot vs bot game.

**Comment by Nostrademous** [comment]
if you have verified that tread swapping works, can you close issue please?

**Comment by pbenologa** [comment]
I havent tested it yet, I only have access to dota over the weekends. I'll close this once i verify

**Comment by Nostrademous** [comment]
I did some testing with this today... made some changes as "using" an ability doesn't mean necessarily that it is fired of that instant... they have a cast point after all, and in some instances we have to move into range. So we can't cycle to INT treads Use Ability and then instantly cycle back. I implemented methods to handle this, but I think a better solution than mine exists, I just don't have brain cycles to currently devote to this.

**Comment by pbenologa** [comment]
> npcBot:Action_UseAbilityOnLocation(tp, dest);
            if tpSwap then 
                npcBot:Action_SwapItems(0, backPackSlot)
            end

So I added a code for using TP. And this line of code is being executed continuosly. Meaning after the bot has finished casting TP it will instantly swap items

**Comment by Nostrademous** [comment]
Following global constants now exist:
ATTRIBUTE_INVALID, ATTRIBUTE_STRENGTH, ATTRIBUTE_AGILITY, and ATTRIBUTE_INTELLECT

**Comment by CsharpNoobDude** [comment]
@Nostrademous Is this already implemented? Because the heroes do not seem to be doing it.

**Comment by Nostrademous** [comment]
The code is written but untested and thus commented out currently, so yes, the bots are not "doing" it. The issue is still to test it and to better protect the action, as remember, tread swapping is an action and will break things like channeling (TP's and spells), break you out of invisibility, etc.

So we need to add some logic to tread usage to make sure we don't swap and break an effect

**Comment by CsharpNoobDude** [comment]
You could check if the bot is channeling, and if not he can do it.
If not npcBot:IsChanneling() then
{
cylceTreadsToAttribute();
}
end;
